docs/adr/0001_cmsis_component_source_structure_recommendations.md

```diff
@@ -0,0 +1,111 @@
+|           +-- TOOLCHAIN_IAR/          # Currently in rtos/source/TARGET_CORTEX/
+|           +-- mbed_lib.json           # Currently in rtos/source/TARGET_CORTEX/
+|       +-- Core_A/
+|           +-- Include/
+|               +-- H                   # Header files currently in cmsis/TARGET_CORTEX_A/
+|           +-- Source/
+|               +-- I                   # Source files currently in cmsis/TARGET_CORTEX_A/
+|       +-- Core/
+|           +-- Include/
+|               +-- J                   # Header files currently in cmsis/TARGET_CORTEX_M/
+```
+
+However, to cope with Mbed-CLI requirement to find directories to process with the target labels mechanism, some directories will be prefixed with `TARGET_` and capitalized. It will therefore look as follows:
+
+```
++-- rtos/CMSIS_5/
+|   +-- README.md
+|   +-- docs/
+|   +-- CMSIS/
+|       +-- include/
+|               +-- mbed_cmsis_conf.h   # Currently in cmsis/
+|       +-- RTE/
+|           +-- include/
+|               +-- RTE_Components.h    # Currently in cmsis/
+|       +-- RTOS2/
+|           +-- Include/
+|               +-- A                   # Content of rtos/source/TARGET_CORTEX/rtx5/Include/
+|               +-- B                   # Header files currently in rtos/source/TARGET_CORTEX/
+|           +-- RTX/
+|               +-- Config/
+|                   +-- C               # Content of rtos/source/TARGET_CORTEX/rtx5/RTX/Config/
+|               +-- Include/
+|                   +-- D               # Content of rtos/source/TARGET_CORTEX/rtx5/RTX/Include/
+|               +-- Include1/
+|                   +-- cmsis_os.h      # Currently in rtos/source/TARGET_CORTEX/rtx4/
+|               +-- Library/
+|                   +-- cmsis_os1.c     # Currently in rtos/source/TARGET_CORTEX/rtx4/
+|               +-- Source/
+|                   +-- E               # Content of rtos/source/TARGET_CORTEX/rtx5/RTX/Source/
+|           +-- Source/
+|               +-- F                   # Content of rtos/source/TARGET_CORTEX/rtx5/Source
+|               +-- G                   # Source files currently in rtos/source/TARGET_CORTEX/
+|           +-- TOOLCHAIN_ARM_MICRO/    # Currently in rtos/source/TARGET_CORTEX/
+|           +-- TOOLCHAIN_ARM_STD/      # Currently in rtos/source/TARGET_CORTEX/
+|           +-- TOOLCHAIN_GCC_ARM/      # Currently in rtos/source/TARGET_CORTEX/
+|           +-- TOOLCHAIN_IAR/          # Currently in rtos/source/TARGET_CORTEX/
+|           +-- mbed_lib.json           # Currently in rtos/source/TARGET_CORTEX/
+|       +-- TARGET_CORTEX_A/
+|           +-- Include/
+|               +-- H                   # Header files currently in cmsis/TARGET_CORTEX_A/
+|           +-- Source/
+|               +-- I                   # Source files currently in cmsis/TARGET_CORTEX_A/
+|       +-- TARGET_CORTEX_M/
+|           +-- Include/
+|               +-- J                   # Header files currently in cmsis/TARGET_CORTEX_M/
+```
+
+## Documentation
```

Comment:
One section to be added, how to import upstream repo. We have a script in tools that fetches the repo and applies our changes on top of it. See `tools/importer` script. Maybe the script should be part of this CMSIS folder?

why it was  written? We did changes to CMSIS (not all were accepted in the upstream repository). We should work on alignment there, Meanwhile however this should work (importing + our local changes).

Reply:
Thanks for this information. I thought it was all done manually. I will update the script now.

Reply:
I was about to write that the tools/importer script should be updated. In fact I think this PR should be the adr + changes to the script.  Then another PR which uses teh script to update the directory structure

Reply:
The script does not need to be updated as it is passed a configuration file to specify where the imported source code lands.
I have updated the configuration file according to the currently proposed adr.
See https://github.com/ARMmbed/mbed-os/pull/13464/commits/db139ab7901ac21a56ecb4bc032c8d31ee3e0b2c

Reply:
There should be a standard way to apply patches to both the OS and libraries brought into it via the tools.

Also +1 to @0xc0170 for recommending getting better alignment with the CMSIS crew. CMSIS and Mbed should work naturally together.

Reply:
Indeed, I meant the conf file not the script. Thanks for updating and adding the note.